rollup/da_syncer/da/calldata_blob_source.go

```diff
@@ -0,0 +1,241 @@
+	revertBatchEventName                      = "RevertBatch"
+	finalizeBatchEventName                    = "FinalizeBatch"
+	commitBatchMethodName                     = "commitBatch"
+	commitBatchWithBlobProofMethodName        = "commitBatchWithBlobProof"
+
+	// the length og method ID at the beginning of transaction data
+	methodIDLength = 4
+)
+
+var (
+	ErrSourceExhausted = errors.New("data source has been exhausted")
+)
+
+type CalldataBlobSource struct {
+	ctx                           context.Context
+	l1Client                      *rollup_sync_service.L1Client
+	blobClient                    blob_client.BlobClient
+	l1height                      uint64
+	scrollChainABI                *abi.ABI
+	l1CommitBatchEventSignature   common.Hash
+	l1RevertBatchEventSignature   common.Hash
+	l1FinalizeBatchEventSignature common.Hash
+	db                            ethdb.Database
+
+	l1Finalized uint64
+}
+
+func NewCalldataBlobSource(ctx context.Context, l1height uint64, l1Client *rollup_sync_service.L1Client, blobClient blob_client.BlobClient, db ethdb.Database) (*CalldataBlobSource, error) {
+	scrollChainABI, err := rollup_sync_service.ScrollChainMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get scroll chain abi: %w", err)
+	}
+	return &CalldataBlobSource{
+		ctx:                           ctx,
+		l1Client:                      l1Client,
+		blobClient:                    blobClient,
+		l1height:                      l1height,
+		scrollChainABI:                scrollChainABI,
+		l1CommitBatchEventSignature:   scrollChainABI.Events[commitBatchEventName].ID,
+		l1RevertBatchEventSignature:   scrollChainABI.Events[revertBatchEventName].ID,
+		l1FinalizeBatchEventSignature: scrollChainABI.Events[finalizeBatchEventName].ID,
+		db:                            db,
+	}, nil
+}
+
+func (ds *CalldataBlobSource) NextData() (Entries, error) {
+	var err error
+	to := ds.l1height + callDataBlobSourceFetchBlockRange
+
+	// If there's not enough finalized blocks to request up to, we need to query finalized block number.
+	// Otherwise, we know that there's more finalized blocks than we want to request up to
+	// -> no need to query finalized block number
+	if to > ds.l1Finalized {
+		ds.l1Finalized, err = ds.l1Client.GetLatestFinalizedBlockNumber()
+		if err != nil {
+			return nil, serrors.NewTemporaryError(fmt.Errorf("failed to query GetLatestFinalizedBlockNumber, error: %v", err))
+		}
+		// make sure we don't request more than finalized blocks
+		to = min(to, ds.l1Finalized)
+	}
+
+	if ds.l1height > to {
+		return nil, ErrSourceExhausted
+	}
+
+	logs, err := ds.l1Client.FetchRollupEventsInRange(ds.l1height, to)
+	if err != nil {
+		return nil, serrors.NewTemporaryError(fmt.Errorf("cannot get events, l1height: %d, error: %v", ds.l1height, err))
+	}
+	da, err := ds.processLogsToDA(logs)
+	if err != nil {
+		return nil, serrors.NewTemporaryError(fmt.Errorf("failed to process logs to DA, error: %v", err))
```

Comment:
why categorize this error as temporary?

Reply:
This is more of a "catch-all" style of wrapping the errors here. There might be dozens of individual errors that are temporary and some might actually be less so. However, a temporary error is handled with a retry and if there's too many retries the pipeline will simply stop. So while it technically is not 100% correct, it also doesn't hurt to have this catch-all temporary error here. 

See here how we handle temporary errors: https://github.com/scroll-tech/go-ethereum/blob/00494d58cf3390cbf909a06ef3ef46b0587d6d90/rollup/da_syncer/syncing_pipeline.go#L180

Reply:
I see. thanks, the original confusion is why errors returned by `processLogsToDA` can be regarded as temporary errors. and then I found that `processLogsToDA` will send rpc queries, so no confusion now.